### PR TITLE
feat: Salesforce RR skip based on a user on a lookup field on an account

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR allows users to add a route under a routing form to look up a user under a custom lookup field under a Salesforce account. If that user is a part of the round robin event, then directly book with that user.

image

- Fixes #XXXX (GitHub issue number)
- Fixes CAL-XXXX (Linear issue number - should be visible at the bottom of the GitHub issue description)

https://www.loom.com/share/db809cd26f56481299eac1a4067c885c?sid=940ebec9-021a-4ee4-876d-3bd86486d235

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Enable Salesforce in an round robin event type
- Go to an event type and choose the option to route to a lookup field under Salesforce
- Enter the field name of the lookup field under a Salesforce account
- Fill out the routing form with an email domain that exists under the account
	- It should route to that user
- Fill out the routing form with an email that doesn't exist under an account
	- It should follow the attribute routing

### CODE REVIEW

File: packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx [hunk 1]
```diff
@@ -41,31 +45,22 @@ import {
...
+  LocalRoute,
+  SerializableRoute,
+  Attribute,
+  EditFormRoute,
+  LocalRouteWithRaqbStates,
+  AttributeRoutingConfig,
+} from "../../types/types";
 import { RouteActionType } from "../../zod";
 
-type FormFieldsQueryBuilderState = {
```
Comment (joeauyeung) [inline packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx:47]: Move types to the types file so they can be used in other components

File: packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx [hunk 2]
```diff
@@ -404,12 +421,25 @@ const Route = ({
...
+          <div className="mt-4 px-2.5">
+            <DynamicAppComponent
+              componentMap={routingFormAppComponents}
+              slug="salesforce"
+              appData={eventTypeRedirectUrlSelectedOption?.eventTypeAppMetadata["salesforce"]}
+              route={route}
+              setAttributeRoutingConfig={setAttributeRoutingConfig}
+            />
+          </div>
+        ) : null}
```
Comment (joeauyeung) [inline packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx:435]: TODO: Make this more dynamic and not hard code "salesforce" here

File: packages/app-store/salesforce/zod.ts [hunk 3]
```diff
@@ -3,6 +3,13 @@ import { z } from "zod";
 import { eventTypeAppCardZod } from "../eventTypeAppCardZod";
 import { SalesforceRecordEnum } from "./lib/recordEnum";
 
+export const routingFormOptions = z
```
Comment (joeauyeung) [inline packages/app-store/salesforce/zod.ts:6]: I think routing form app data should be separate from the other app data. What do you think?

Comment (hariombalhara) [reply packages/app-store/salesforce/zod.ts:6]: Yeah makes sense

File: packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx [hunk 4]
```diff
@@ -196,16 +195,34 @@ const buildEventsData = ({
...
+        const appMetadata = eventTypeAppMetadataParse.data;
+
+        if (appMetadata) {
+          for (const appSlug of Object.keys(appMetadata)) {
+            if (enabledAppSlugs.includes(appSlug)) {
+              eventTypeAppMetadata[appSlug] = appMetadata[appSlug as keyof typeof appMetadata];
+            }
+          }
+        }
+      }
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx:212]: I know it is in draft but still leaving code feedback.


Let's abstract it out

File: apps/web/lib/team/[slug]/[type]/getServerSideProps.tsx [hunk 5]
```diff
@@ -112,21 +118,131 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
...
+
+  return null;
+}
+
+async function handleRoutingFormOption(query: ParsedUrlQuery, eventTypeId: number) {
+  const nullReturnValue = { email: null, skipContactOwner: false };
+
+  if (typeof query.email !== "string") return nullReturnValue;
+
+  const routingFormQuery = await prisma.app_RoutingForms_Form.findFirst({
```
Comment (joeauyeung) [inline apps/web/lib/team/[slug]/[type]/getServerSideProps.tsx:167]: The app specific options live in the routing form data. I didn't want to leak any Salesforce specific information in the URL

File: apps/web/lib/team/[slug]/[type]/getServerSideProps.tsx [hunk 6]
```diff
@@ -112,21 +118,131 @@ export const getServerSideProps = async (context: GetServerSidePropsContext) =>
...
+
+    if (enabledAppSlugs.includes(appSlug)) {
+      enabledRoutingFormApp = appSlug;
+      break;
+    }
+  }
+
+  if (!enabledRoutingFormApp) return nullReturnValue;
+
+  const appHandler = appBookingFormHandler[enabledRoutingFormApp];
```
Comment (joeauyeung) [inline apps/web/lib/team/[slug]/[type]/getServerSideProps.tsx:220]: Setting us up for when we extend this functionality to other CRMs

File: packages/app-store/routing-forms/appBookingFormHandler.ts
Comment (joeauyeung) [inline packages/app-store/routing-forms/appBookingFormHandler.ts:1]: This is where each app specific routing form booking form logic will live.

File: packages/app-store/routing-forms/appComponents.ts
Comment (joeauyeung) [inline packages/app-store/routing-forms/appComponents.ts:1]: Same with the components that are rendered on the routing form route builder.

File: packages/app-store/routing-forms/appDataSchemas.ts
Comment (joeauyeung) [inline packages/app-store/routing-forms/appDataSchemas.ts:1]: Also any metadata associated with any apps enabled for routing forms.

File: packages/app-store/routing-forms/components/DynamicAppComponent.tsx
Comment (joeauyeung) [inline packages/app-store/routing-forms/components/DynamicAppComponent.tsx:1]: This component is used to dynamically render app specific options on the route builder.

File: packages/app-store/routing-forms/lib/getEventTypeAppMetadata.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import { EventTypeMetaDataSchema } from "@calcom/prisma/zod-utils";
+
+import { enabledAppSlugs } from "./enabledApps";
+import type { Prisma } from ".prisma/client";
+
+const getEventTypeAppMetadata = (metadata: Prisma.JsonValue) => {
```
Comment (joeauyeung) [inline packages/app-store/routing-forms/lib/getEventTypeAppMetadata.ts:6]: This is used to get the app metadata from the event type for apps that are enabled for routing forms.

File: packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx [hunk 8]
```diff
@@ -404,11 +380,17 @@ const Route = ({
...
+        "salesforce" in eventTypeRedirectUrlSelectedOption.eventTypeAppMetadata ? (
+          <div className="mt-4 px-2.5">
+            <DynamicAppComponent
+              componentMap={routingFormAppComponents}
+              slug="salesforce"
+              appData={eventTypeRedirectUrlSelectedOption?.eventTypeAppMetadata["salesforce"]}
+              route={route}
+              setAttributeRoutingConfig={setAttributeRoutingConfig}
+            />
+          </div>
```
Comment (joeauyeung) [inline packages/app-store/routing-forms/pages/route-builder/[...appPages].tsx:393]: Replace the previous switch with dynamically rendered app options.

File: packages/app-store/routing-forms/pages/routing-link/getUrlSearchParamsToForward.ts [hunk 9]
```diff
@@ -80,15 +81,34 @@ export function getUrlSearchParamsToForward({
...
+      if (key === "skipContactOwner" && attributeRoutingConfig[key]) {
+        attributeRoutingConfigParams["cal.skipContactOwner"] = "true";
+      }
+
+      // TODO: How do we move this logic to their respective app packages
+      if (key === "salesforce") {
+        const salesforceData = attributeRoutingConfig[key];
+
+        if (salesforceData?.rrSkipToAccountLookupField && salesforceData.rrSKipToAccountLookupFieldName) {
+          attributeRoutingConfigParams["cal.salesforce.rrSkipToAccountLookupField"] = "true";
```
Comment (joeauyeung) [inline packages/app-store/routing-forms/pages/routing-link/getUrlSearchParamsToForward.ts:97]: Only want to send to the URL that this option is enabled. Don't want to send the specific field name as that is unique to each Salesforce instance.

File: packages/app-store/salesforce/components/RoutingFormOptions.tsx
Comment (joeauyeung) [inline packages/app-store/salesforce/components/RoutingFormOptions.tsx:1]: This is the Salesforce specific app options. Notice how this lives under `/salesforce` folder and not under the routing form folder.

File: packages/app-store/salesforce/components/RoutingFormOptions.tsx [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    } else {
+      return salesforceRoutingOptions.find((option) => option.value === SalesforceRoutingConfig.ROUTE_OWNER);
+    }
+  };
+
+  const [selectedRoutingOption, setSelectedRoutingOption] = useState(getInitialOption());
+
+  useEffect(() => {
+    setSelectedRoutingOption(getInitialOption());
+  }, [route]);
```
Comment (joeauyeung) [inline packages/app-store/salesforce/components/RoutingFormOptions.tsx:57]: The route will change when we change the event type to redirect to.

File: packages/app-store/salesforce/lib/CrmService.ts [hunk 11]
```diff
@@ -864,4 +863,45 @@ export default class SalesforceCRMService implements CRM {
...
+    const conn = await this.conn;
+
+    // Ensure the field exists on the record
+    const existingFields = await this.ensureFieldsExistOnObject([fieldName], salesforceObject);
+
+    if (!existingFields.length) return;
+
+    const lookupField = existingFields[0];
+
+    if (salesforceObject === SalesforceRecordEnum.ACCOUNT) {
```
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/CrmService.ts:881]: Only have functionality for account objects right now. We can always add logic for other types of records later.

File: packages/app-store/salesforce/lib/routingFormBookingFormHandler.ts
Comment (joeauyeung) [inline packages/app-store/salesforce/lib/routingFormBookingFormHandler.ts:1]: Logic specific to the booking page. Also lives under the `/salesforce` folder.

**Review by emrysal — DISMISSED** [review]
Happy with this, thanks for your time to walk me through it; lgtm 🚀

**Review by emrysal — APPROVED** [review]

**Pull request merged**